File: 03 C C++ Python R and Mojo review/03-3 R Language Review/01 Pointers/04.R
# Simulating NULL pointer behavior in R

ptr <- NULL  # Initialize ptr as NULL
cat("1. Initial ptr value:", ptr, "\n")

# Check for NULL before using
if (is.null(ptr)) {
    cat("2. ptr is NULL, cannot dereference\n")
}

# Simulate memory allocation
ptr <- 42  # Assigning a value to ptr (simulating allocation)
cat("4. After allocation, ptr value:", ptr, "\n")

# Safe to use ptr after NULL check
if (!is.null(ptr)) {
    cat("5. Value at ptr:", ptr, "\n")
}

# Simulate freeing memory (set to NULL after use)
ptr <- NULL
cat("6. After free, ptr value:", ptr, "\n")

# Safety check after 'free' (set to NULL)
if (is.null(ptr)) {
    cat("7. ptr is NULL, safely avoided use after free\n")
}

File: 03 C C++ Python R and Mojo review/03-2 Python Review/01 Pointers/04.py
# Simulating NULL pointer behavior in Python using None

ptr = None  # Initializing to None (similar to NULL in C)
print(f"1. Initial ptr value: {ptr}")

# Check for None (NULL) before using
if ptr is None:
    print("2. ptr is None, cannot dereference")

# Simulate memory allocation
ptr = 42  # Allocating a value to ptr
print(f"4. After allocation, ptr value: {ptr}")

# Safe to use ptr after checking for None
if ptr is not None:
    print(f"5. Value at ptr: {ptr}")

# Simulate freeing memory (set to None after use)
ptr = None
print(f"6. After free, ptr value: {ptr}")

# Safety check after 'free' (set to None)
if ptr is None:
    print("7. ptr is None, safely avoided use after free")